sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuAvroScan.scala

```diff
@@ -151,92 +151,276 @@ case class GpuAvroPartitionReaderFactory(
 
   override def buildColumnarReader(partFile: PartitionedFile): PartitionReader[ColumnarBatch] = {
     val conf = broadcastedConf.value.value
-    val blockMeta = GpuAvroFileFilterHandler(sqlConf, broadcastedConf,
-      ignoreExtension, broadcastedConf.value.value).filterBlocks(partFile)
-    val reader = new PartitionReaderWithBytesRead(new AvroPartitionReader(conf, partFile, blockMeta,
-      readDataSchema, debugDumpPrefix, maxReadBatchSizeRows,
+    val blockMeta = AvroFileFilterHandler(conf, options).filterBlocks(partFile)
+    val reader = new PartitionReaderWithBytesRead(new GpuAvroPartitionReader(conf, partFile,
+      blockMeta, readDataSchema, debugDumpPrefix, maxReadBatchSizeRows,
       maxReadBatchSizeBytes, metrics))
     ColumnarPartitionReaderWithPartitionValues.newReader(partFile, reader, partitionSchema)
   }
 }
 
 /**
- * A tool to filter Avro blocks
- *
- * @param sqlConf         SQLConf
- * @param broadcastedConf the Hadoop configuration
+ * The multi-file partition reader factory for cloud or coalescing reading of avro file format.
  */
-private case class GpuAvroFileFilterHandler(
+case class GpuAvroMultiFilePartitionReaderFactory(
     @transient sqlConf: SQLConf,
     broadcastedConf: Broadcast[SerializableConfiguration],
-    ignoreExtension: Boolean,
-    hadoopConf: Configuration) extends Arm with Logging {
+    dataSchema: StructType,
+    readDataSchema: StructType,
+    partitionSchema: StructType,
+    filters: Array[Filter],
+    @transient rapidsConf: RapidsConf,
+    options: AvroOptions,
+    metrics: Map[String, GpuMetric],
+    queryUsesInputFile: Boolean)
+  extends MultiFilePartitionReaderFactoryBase(sqlConf, broadcastedConf, rapidsConf) {
 
-  def filterBlocks(partFile: PartitionedFile): AvroBlockMeta = {
+  private val debugDumpPrefix = rapidsConf.avroDebugDumpPrefix
+  private val ignoreMissingFiles = sqlConf.ignoreMissingFiles
+  private val ignoreCorruptFiles = sqlConf.ignoreCorruptFiles
+
+  private val numThreads = rapidsConf.avroMultiThreadReadNumThreads
+  private val maxNumFileProcessed = rapidsConf.maxNumAvroFilesParallel
+
+  // Disable coalescing reading until it is supported.
+  override val canUseCoalesceFilesReader: Boolean = false
```

Comment:
Could we enable this and force the reader to multithreaded and add the unit tests in GpuReaderSuite.scala

Reply:
Instead I changed the tests a little to cover the cases for the avro reader type check.